File: src/main/java/net/yogstation/yogbot/util/Result.java
package net.yogstation.yogbot.util;

public class Result<T, E> {
	private final T value;
	private final E error;

	private Result(T success, E error) {
		this.value = success;
		this.error = error;
	}

	public static<T, E> Result<T, E> success(T success) {
		return new Result<>(success, null);
	}

	public static<T, E> Result<T, E> error(E error) {
		return new Result<>(null, error);
	}

	public boolean hasError() {
		return error != null;
	}

	public T getValue() {
		return value;
	}

	public E getError() {
		return error;
	}
}

File: src/main/java/net/yogstation/yogbot/util/Result.kt
package net.yogstation.yogbot.util

class Result<T, E> private constructor(val value: T?, val error: E?) {

	fun hasError(): Boolean {
		return error != null
	}

	fun hasValue(): Boolean {
		return value != null
	}

	companion object {

		fun <T, E> success(success: T): Result<T, E?> {
			return Result(success, null)
		}

		fun <T, E> error(error: E): Result<T?, E> {
			return Result(null, error)
		}
	}
}
